---
title: "HumanTranscriptomeCompendium -- a cloud-resident collection of sequenced human transcriptomes"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{HumanTranscriptomeCompendium -- a cloud-resident collection of sequenced human transcriptomes}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

```{r setup,echo=FALSE,results="hide"}
suppressPackageStartupMessages({
suppressMessages({
library(BiocStyle)
library(HumanTranscriptomeCompendium)
library(beeswarm)
library(SummarizedExperiment)
library(DT)
})
})
```

# Introduction

Comprehensive archiving of genome-scale sequencing experiments
is valuable for substantive and methodological progress in
multiple domains.

The `r Biocpkg("HumanTranscriptomeCompendium")` package provides functions for interacting
with quantifications and metadata for over 180000 sequenced human
transcriptomes.

# Access to gene-level quantifications

`r Biocpkg("BiocFileCache")` is used to manage access
to a modest collection of metadata about compendium
contents.  By default, `htx_load` will
load the cache and establish a connection to
remote HDF5 representation of quantifications.
The numerical data is lodged in an instance of
the HDF Scalable Data Service, at `http://hsdshdflab.hdfgroup.org`.

```{r lklo}
library(HumanTranscriptomeCompendium)
genelev = htx_load()
genelev
assay(genelev)
```

## Identifying single-cell RNA-seq studies

We use crude pattern-matching in the study titles
to identify single cell RNA-seq experiments 
```{r lksi}
sing = grep("single.cell", genelev$study_title, 
   ignore.case=TRUE)
length(sing)
```

Now we will determine which studies are involved.
We will check out the titles of the single-cell
studies to assess the specificity of this approach.

```{r chkspec}
sa = genelev$study_accession[sing]
sing2 = sing[-which(duplicated(sa))]
length(sing2)
datatable(as.data.frame(colData(genelev[,sing2])),
   options=list(lengthMenu=c(3,5,10,50,100)))
```

## Collecting bulk RNA-seq samples on a disease of interest: glioblastoma


```{r lkchar}
bulk = genelev[,-sing]
kpglio = grep("glioblastoma", bulk$study_title, 
  ignore.case=TRUE)
glioGene = bulk[,kpglio]
glioGene
```

To acquire numerical values, `as.matrix(assay())` is needed.
```{r lkmat}
beeswarm(as.matrix(assay(
   glioGene["ENSG00000138413.13",1:100])), pwcol=as.numeric(factor(glioGene$study_title[1:100])), ylab="IDH1 expression")
legend(.6, 15000, legend=unique(glioGene$study_accession[1:100]),
   col=1:2, pch=c(1,1))
```

# Access to transcript-level quantifications

By setting `genesOnly` to FALSE in `htx_load`,
we get a transcript-level version of the compendium.
Note that the number of samples in this version exceeds
that of the gene version by two.  There are two
unintended columns in the underlying HDF Cloud
array, with names 'X0' and 'X0.1', that should
be ignored.

```{r dotx}
txlev = htx_load(genesOnly=FALSE)
txlev
```

# Resources of the HumanTranscriptomeCompendium package

The primary purposes of the HumanTranscriptomeCompendium
package are

- providing programmatic acccess to the remote HDF5
representation of the compendium
- providing access to fundamental metadata about the
contents of the compendium
- providing an approach to verifying the correctness
of the representation in the package.

We will address these in turn.

## Access to the quantifications

### `htx_load`

`htx_load` has three arguments: `remotePath`, `cache`, and `genesOnly`.

`genesOnly` defaults to TRUE.  If it is TRUE, the HDF array that
will be used consists of gene-level quantifications; otherwise
the array in use will consist of transcript-level quantifications
based on the Gencode V27 models.

`remotePath` is the path to an RDS-formatted RangedSummarizedExperiment
instance that has been prepared to include a DelayedArray
reference to the HSDS representation of the quantifications.  The
specific reference used depends on the setting of `genesOnly`.
The default value currently references an AWS S3 bucket to 
retrieve the RDS.

`cache` is an instance of `BiocFileCache`, where the RDS
will be stored and retrieved as needed.

A typical use is `htx = htx_load()` which efficiently sets
up `htx` to give access to gene-level quantifications.
After such a command is issued, `assay(htx[G, S])` is the
DelayedMatrix for features `G` on samples `S`.  If `G` or `S`
are too long, the HSDS may return an error.  Systematic
chunking of large requests is a topic of future development.

### `htx_query`

`htx_query` has one mandatory argument,
`study_accessions`.  This function uses `htx_load` to prepare a SummarizedExperiment
with DelayedArray assay data,
with samples limited to those in the studies listed in the character vector
argument `study_accessions`.  Optional arguments to this function
are passed to `htx_load`.

### `htx_app`

`htx_app` has no arguments.  It fires up a shiny app that lists studies by
size, study accession number, and study title.  User can search titles
using regular expressions, and can ask for retrieval of multiple studies.
The studies are returned in a SummarizedExperiment.  This is for use in R.
A more advanced query/retrieval app is prototyped at vjcitn.shinyapps.io/cancer9k.
The cancer9k app provides a 'search engine'-like capability over a richer
collection of sample-level attributes.  See the package at vjcitn/htxapp
for the sources related to cancer9k.

 [1] "addRD"                                
 [2] "bigrnaFiles"                          
 [3] "experTable"                           
 [4] "htx_app"                              
 [7] "HumanTranscriptomeCompendium.colnames"
 [8] "procExpToGene"                        
 [9] "sampleAtts"                           
[10] "studTable"                            
[11] "tx2gene_gencode27"                    
[12] "uniqueAcc_120518"                     
